### PR TITLE
SRTO_OUTPACEMODE option for Output Pace Control

### PR DESCRIPTION
**Pull request by jeandube**
SRTO_OUTPACEMODE option can be use to implement custom sender output bit rate. This option provides all the implicit output pace modes defined by the options SRTO_MAXBW, SRTO_INPUTBW, and SRTO_OHEADBW, and add a new mode combining the internally sampled input rate and the configured input rate to adjust the output rate. Other custom modes can be added taking other factors into account such as he size of the sender buffer, the size of the lost packets list, etc... Documentation has been updated and enhanced. The GET-only option SRTO_SMPINBW has also been added to get the internally sampled input bitrate but this value should probably be added in the stats also.

### CODE REVIEW

File: srtcore/core.cpp [hunk 1]
```diff
@@ -576,6 +579,37 @@ void CUDT::setOpt(SRT_SOCKOPT optName, const void* optval, int optlen)
...
+            break;
+        case SRT_OPM_INBWADJ:
+            //m_llInputBW is or will be set.
+            m_llMaxBW = 0LL;
+            break;
+        default:
+            throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+            break;
+        }
+        m_iOutPaceMode = *(SRT_OUTPACEMODE *)optval;
```
Comment (maxsharabayko) [inline srtcore/core.cpp:608]:
```suggestion
        m_iOutPaceMode =cast_optval<SRT_OUTPACEMODE>(optval, optlen);
```

File: srtcore/core.cpp [hunk 2]
```diff
@@ -1092,11 +1126,58 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
...
+               *(int *)optval = SRT_OPM_CAPPED;
+           }else if (m_llMaxBW == 0) {
+               //Automatic based on set or sampled inputBW
+               if(m_llInputBW > 0)  *(int *)optval = SRT_OPM_INBWSET;
+               else
+               if(m_llInputBW == 0) *(int *)optval = SRT_OPM_SMPINBW;
+           }
+           //SRT_OPM_UNSET returned on invalid/ambiguous setting
+       }
+       break;
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1146]: Formatting.

```suggestion
    case SRTO_OUTPACEMODE:
        *(int*)optval = m_iOutPaceMode;
        optlen        = sizeof(int);
        if (m_iOutPaceMode == SRT_OPM_UNSET)
        {
            // unset return the equivalence based on pre-outpacemode settings
            if (m_llMaxBW == -1 || m_llMaxBW == BW_INFINITE)
            {
                *(int*)optval = SRT_OPM_UNTAMED;
            }
            else if (m_llMaxBW > 0)
            {
                *(int*)optval = SRT_OPM_CAPPED;
            }
            else if (m_llMaxBW == 0)
            {
                // Automatic based on set or sampled inputBW
                if (m_llInputBW > 0)
                    *(int*)optval = SRT_OPM_INBWSET;
                else if (m_llInputBW == 0)
                    *(int*)optval = SRT_OPM_SMPINBW;
            }
            // SRT_OPM_UNSET returned on invalid/ambiguous setting
        }
        break;
```

File: srtcore/core.cpp [hunk 3]
```diff
@@ -1092,11 +1126,58 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
...
+	      //return sampled internally measured input bw
+	      uint64_t llSmpInputBW;
+	      llSmpInputBW = m_pSndBuffer->getInputRate(); //Auto input rate
+	      if( m_pSndBuffer->getInRatePeriod() != 0) { //sampling active
+		  *(int64_t*)optval =llSmpInputBW;
+	      }
+	  }
+      }
+      optlen = sizeof(int64_t);
+      break;
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1174]: Formatting.

```suggestion
    case SRTO_SMPINBW:
        *(int64_t*)optval = 0LL;
        if ((m_iOutPaceMode == SRT_OPM_SMPINBW) || (m_iOutPaceMode == SRT_OPM_INBWADJ) ||
            ((m_iOutPaceMode == SRT_OPM_UNSET) && (m_llInputBW == 0LL)))
        {
            if (m_pSndBuffer)
            {
                // return sampled internally measured input bw
                uint64_t llSmpInputBW;
                llSmpInputBW = m_pSndBuffer->getInputRate(); // Auto input rate
                if (m_pSndBuffer->getInRatePeriod() != 0)
                { // sampling active
                    *(int64_t*)optval = llSmpInputBW;
                }
            }
        }
        optlen = sizeof(int64_t);
        break;
```

File: srtcore/core.cpp [hunk 4]
```diff
@@ -1092,11 +1126,58 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
...
+
+   case SRTO_OHEADBW:
+      *(int *)optval = m_iOverheadBW;
+      optlen = sizeof(int);
+      break;
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1179]: Formatting: 4 space alignment.

```suggestion
    case SRTO_OHEADBW:
        *(int *)optval = m_iOverheadBW;
        optlen = sizeof(int);
        break;
```

File: srtcore/core.cpp [hunk 5]
```diff
@@ -7673,19 +7762,21 @@ bool CUDT::updateCC(ETransmissionEvent evt, const EventVariant arg)
     {
         // Specific part done when MaxBW is set to 0 (auto) and InputBW is 0.
         // This requests internal input rate sampling.
-        if (m_llMaxBW == 0 && m_llInputBW == 0)
+        if (((m_llMaxBW == 0 && m_llInputBW == 0) || (m_iOutPaceMode == SRT_OPM_SMPINBW))
+        // or we adjust InputBW with sampled internal input rate when higher to adjust to encoder bitrate overshoot
+        ||  (m_iOutPaceMode == SRT_OPM_INBWADJ))
```
Comment (maxsharabayko) [inline srtcore/core.cpp:7767]:
```suggestion
        if (((m_llMaxBW == 0 && m_llInputBW == 0) || (m_iOutPaceMode == SRT_OPM_SMPINBW))
            // or we adjust InputBW with sampled internal input rate when higher to adjust to encoder bitrate overshoot
            ||  (m_iOutPaceMode == SRT_OPM_INBWADJ))
```

File: srtcore/srt.h [hunk 6]
```diff
@@ -319,6 +320,16 @@ static const int SRT_LIVE_MAX_PLSIZE = 1456; // MTU(1500) - UDP.hdr(28) - SRT.hd
...
 
+typedef enum SRT_OUTPACEMODE
+{
+    SRT_OPM_UNSET,      //unset, API/ABI backward compatible mode
+    SRT_OPM_UNTAMED,    //unlimited: SRTO_MAXBW == -1)
+    SRT_OPM_CAPPED,     //capped by (SRTO_MAXBW > )
+    SRT_OPM_SMPINBW,    //based on internally sampled input rate and configured overhead: maxoutBW = sampled-input-bw * (1+SRTO_OVERHEAD/100)
+    SRT_OPM_INBWSET,    //based on configured input rate and overhead (SRTO_INPUTBW): maxoutBW = SRTO_INPUTBW * (1+SRTO_OVERHEAD/100)
+    SRT_OPM_INBWADJ     //based on configured input rate and overhead, adjusted to internally sampled input rate when overshoot configured value
+}SRT_OUTPACEMODE;
```
Comment (maxsharabayko) [inline srtcore/srt.h:331]:
```suggestion
} SRT_OUTPACEMODE;
```

File: srtcore/queue.cpp [hunk 7]
```diff
@@ -604,8 +604,29 @@ void *CSndQueue::worker(void *param)
             self->m_WorkerStats.lNotReadyPop++;
 #endif /* SRT_DEBUG_SNDQ_HIGHRATE */
         }
-
+#if 0 //>>duB:debug OutPace
```
Comment (maxsharabayko) [inline srtcore/queue.cpp:607]: This should allow enabling this code with `cmake . -DENFORCE_SRT_DEBUG_OUTPACE=1`
```suggestion
#if defined(SRT_DEBUG_OUTPACE)
```

File: srtcore/queue.cpp [hunk 8]
```diff
@@ -604,8 +604,29 @@ void *CSndQueue::worker(void *param)
...
+                    diff_ots = ts - prev_ots;
+                }else{
+                    ots_str = " ots: -";
+                    diff_ots = prev_ots - ts;
+                }
+                uint64_t ointerval = self->m_pSndUList->m_pHeap[0]->m_pUDT->m_ullInterval_tk;
+                LOGC(mglog.Note, log << self->CONID() << "chn:SENDING SIZE " << pkt.getLength() << " SEQ: " << pkt.getSeqNo() << " its: " << its-prev_its << ots_str << diff_ots <<"/" << ointerval);
+                if(ts > 1)prev_ots = ts;
+                prev_its = its;
+        }
```
Comment (maxsharabayko) [inline srtcore/queue.cpp:626]: Formatting: 8 spaces replaced with 4 spaces.
```suggestion
        {
            static uint64_t prev_ots = 0;
            uint64_t diff_ots;
            static int prev_its = 0;
            int its = pkt.getMsgTimeStamp();
            char *ots_str = " ots: ";
            if ((ts > prev_ots)) {
                ots_str = " ots: ";
                diff_ots = ts - prev_ots;
            } else {
                ots_str = " ots: -";
                diff_ots = prev_ots - ts;
            }
            const uint64_t ointerval = self->m_pSndUList->m_pHeap[0]->m_pUDT->m_ullInterval_tk;
            LOGC(mglog.Note, log << self->CONID() << "chn:SENDING SIZE " << pkt.getLength() << " SEQ: " << pkt.getSeqNo() << " its: " << its-prev_its << ots_str << diff_ots <<"/" << ointerval);
            if (ts > 1)
                prev_ots = ts;
            prev_its = its;
        }
```

**Comment by mbakholdina** [comment]
@jeandube, @maxsharabayko  

Hi Jean, 
I've created a PR #1533 adding the documentation to the current PR with appropriate suggested names for the option and modes so that they are aligned with the RFC. As a result,
- [ ] I suggest changing socket option name from `SRTO_OUTPACEMODE` to `SRTO_PACINGMODE`. The module will be name Pacing Control module instead of Output pace control module and I will change it in RFC as well,
- [ ] I suggest changing modes names to `SRT_PACING_...`, please check my PR for the list of exact names,
- [ ] I've added a very important note in the description of `SRTO_PACINGMODE` option assuming that we need to have additional checks and throw an error in case of incorrect configuring of SRTO_PACINGMODE and a combination of SRTO_MAXBW, SRTO_INPUTBW and SRTO_OHEADBW options. As I can see from the code, we do not do those checks currently and there is a chance for user to make a mistake. Let's discuss which error can be thrown here and where should be add these checks: `setOpt` function? @maxsharabayko What do you recommend?

That's my note:
It's important to note that options [SRTO_MAXBW](#SRTO_MAXBW), [SRTO_INPUTBW](#SRTO_INPUTBW) and [SRT_OHEADBW](#SRT_OHEADBW) should be set first, right before setting the `SRTO_PACINGMODE` option. The order is important. In case the combination of `SRTO_MAXBW`, `SRTO_INPUTBW` and `SRTO_OHEADBW` options is configured wrong for a particular pacing control mode, SRT will throw an error (TODO: discuss what kind of error and put this information here).

- [ ] Please in the code do appropriate renaming of things like `m_iOutPaceMode` to `m_iPacingMode`, etc., so that it corresponds to our new notations
- [ ] Please delete your changes related to describing new socket option in `API.md`, we will merge your PR together with mine #1533 introducing improved version of documentation
- [ ] Please remove `docs/output-pace-control.md` file from the PR, we will have a link to RFC from the socket description as discussed

Please let me know if you have any concerns/corrections/suggestions to all mentioned above.

**Comment by oviano** [comment]
So just on this, I was reading the SRT whitepaper and it states "Another change made in SRT version 1.3 was to add a configuration option called OUTPUTPACEMODE"....just so I'm clear, I assume that was presumably pulled and is now intended for version 1.5 at some point in the future, correct?

**Comment by jeandube** [comment]
@oviano I tried to introduce this feature in 1.4.2 but the focus was on on other features. After some reviews changes were proposed in the form of another PR. The principle is the same but with some naming changes. I am not sure if it will find its way in 1.4.3 or 1.5.0. I don't even know what the next release will be.

**Comment by oviano** [comment]
Right ok thanks, so the whitepaper should be ignored then it would seem.

I’ve clearly been using SRT wrong until now by relying on the input rate to be automatically calculated, since specifying number actual encoder bitrate everything is much more stable and consistent.

On the subject of encoder overshoot; if you specify the input bitrate INPUTBW is there any scenario where you *wouldn’t* want it to use the calculated input rate of it was higher? Seems like it should work this way on any case, if you are specifying the input manually...

**Comment by jeandube** [comment]
@oviano I like the idea of SRTO_MINBW. But unlike SRTO_MAXBW, to replace SRTO_INPUTBW, SRTO_MINBW must be a post-binding option settable at all time. I am using SRTO_INPUTBW when I adjust the encoder's bit rate based on the SRT stats. For your other question I see no reason not to use the measured input bandwidth when higher then configured INPUTBW (overshoot) but I encountered this with adverse effect on decoder only for very low bit rate (~32kbps). I am happy to see other than me needs that feature.

**Comment by oviano** [comment]
So I was thinking SRTO_MINBW would be an additional option and wouldn’t replace INPUTBW.

Ie If you are guaranteeing a fixed bitrate, you use INPUTBW as now. So nothing changes.

But if you want this new behaviour then you specify MINBW instead, meaning “this is my encoder bitrate, but I’m happy for SRT to go above this if the calculated value is higher.”

**Comment by maxsharabayko** [comment]
There was some rework in #1541, but it ended up with a similar conclusion, that `SRTO_MININPUTBW` would be an easier and a more straightforward approach. Even though it adds another integer to be stored in memory.

**Comment by maxsharabayko** [comment]
Closing this PR.
PR #1791 introduced `SRTO_MININPUTBW` option covering the required functionality.
The functionality with pacing mode socket option `SRTO_PACINGMODE` remains in PR #1541, although is likely to be reconsidered.

**Pull request closed without merging**